lineapy/graph_reader/graph_refactorer.py

```diff
@@ -0,0 +1,734 @@
+    ) -> str:
+        """
+        Return a codeblock to call the function with return variables of the graph segment
+
+        :param int indentation: indentation size
+        :param bool keep_lineapy_save: whether do lineapy.save() after execution
+        :param Optional[str] result_placeholder: if not null, append the return result to the result_placeholder
+        """
+        indentation_block = " " * indentation
+        return_string = ", ".join(self.return_variables)
+        args_string = ", ".join(sorted([v for v in self.input_variables]))
+
+        codeblock = f"{indentation_block}{return_string} = get_{self.artifact_safename}({args_string})"
+        if (
+            keep_lineapy_save
+            and self.segment_type == GraphSegmentType.ARTIFACT
+        ):
+            codeblock += f"""\n{indentation_block}lineapy.save({self.return_variables[0]}, "{self.artifact_name}")"""
+        if result_placeholder is not None:
+            codeblock += f"""\n{indentation_block}{result_placeholder}.append(copy.deepcopy({self.return_variables[0]}))"""
+
+        return codeblock
+
+    def get_import_block(self, indentation=0) -> str:
+        """
+        Return a code block for import statement of the graph segment
+        """
+        indentation_block = " " * indentation
+        import_code = get_source_code_from_graph(self.graph_segment).__str__()
+        import_codeblock = "\n".join(
+            [
+                f"{indentation_block}{line}"
+                for line in import_code.split("\n")
+                if len(line.strip(" ")) > 0
+            ]
+        )
+        if len(import_codeblock) > 0:
+            import_codeblock += "\n" * 2
+        return import_codeblock
+
+
+@dataclass
+class SessionArtifacts:
+    """
+    Refactor a given session graph for use in a downstream task (e.g., pipeline building).
+    """
+
+    session_id: LineaID
+    graph: Graph
+    db: RelationalLineaDB
+    nx_graph: nx.DiGraph
+    artifact_segments: List[GraphSegment]
+    import_segment: GraphSegment
+    artifact_list: List[LineaArtifact]
+    artifact_ordering: List
+    artifact_dict: Dict[LineaID, Optional[str]]
+    variable_dict: Dict[LineaID, Set[str]]
+    import_dict: Dict[LineaID, Set[str]]
+    node_context: Dict[LineaID, Dict[str, Any]]
+    input_parameters: List[str]
+    input_parameters_node: Dict[str, LineaID]
+
+    def __init__(self, artifacts: List[LineaArtifact]) -> None:
+        self.artifact_list = artifacts
+        self.session_id = artifacts[0]._session_id
+        self.db = artifacts[0].db
+        self.graph = artifacts[0]._get_graph()
+        self.nx_graph = self.graph.nx_graph
+        self.artifact_segments = []
+        self.node_context = OrderedDict()
+        self.input_parameters = []
+
+        self._update_node_context()
+        self._slice_session_artifacts()
+
+    def _is_import_node(self, node_id: LineaID) -> bool:
```

Comment:
This function could potentially be lifted out to a util file for `graph_reader`, since (1) it's pretty standalone, and (2) other graph readers may also use it.

Reply:
done